lib/manager/components/transform/NormalizeField.js

```diff
@@ -0,0 +1,223 @@
+import Select from 'react-select'
+
+import {getGtfsSpec} from '../../../common/util/config'
+
+import type {
+  Feed,
+  FeedTransformation as FeedTransformationType,
+  GtfsSpecTable,
+  NormalizeFieldFields,
+  ReactSelectOption
+} from '../../../types'
+
+type Props = {
+  feedSource: Feed,
+  index: number,
+  onSave: (NormalizeFieldFields, number) => void,
+  transformation: FeedTransformationType
+}
+
+function getTransformationFields (transformation: FeedTransformationType): NormalizeFieldFields {
+  const {
+    capitalize,
+    capitalizeExceptions = '',
+    fieldName = '',
+    performSubstitutions,
+    substitutions = ''
+  } = transformation
+  return {
+    capitalize,
+    capitalizeExceptions,
+    fieldName,
+    performSubstitutions,
+    substitutions
+  }
+}
+
+/**
+ * Obtain the list of validation issues for the given transformation.
+ */
+export function getNormalizeFieldsValidationIssues (transformation: FeedTransformationType, feedSource: Feed): Array<string> {
+  const issues = []
+  const { fieldName, performSubstitutions, substitutions } = transformation
+
+  // fieldName must be defined.
+  if (!fieldName || fieldName.length === 0) {
+    issues.push('Field to normalize must be defined.')
+  }
+
+  // If substitutions are enabled, then substitutions must be defined.
+  if (performSubstitutions && (!substitutions || substitutions.length === 0)) {
+    issues.push('Substitutions must be defined when enabling substitutions.')
+  }
+
+  return issues
+}
+
+/**
+ * Component that renders input fields for the NoramlizeFieldTransformation.
+ */
+export default class NormalizeField extends Component<Props, NormalizeFieldFields> {
+  state = {
+    capitalize: true,
+    capitalizeExceptions: '',
+    fieldName: '',
+    performSubstitutions: true,
+    substitutions: ''
+  }
+
+  constructor (props: Props) {
+    super(props)
+    this.state = getTransformationFields(props.transformation)
+  }
+
+  _onChangeFieldToNormalize = (fieldOption: ReactSelectOption) => {
+    this.setState({fieldName: fieldOption.value})
+  }
+
+  _onChangeCapitalizeValues = (evt: SyntheticInputEvent<HTMLInputElement>) => {
+    this.setState({capitalize: evt.target.checked})
+  }
+
+  _onChangeCapitalizeExceptions = (evt: SyntheticInputEvent<HTMLInputElement>) => {
+    this.setState({capitalizeExceptions: evt.target.value})
+  }
+
+  _onChangePerformSubstitutions = (evt: SyntheticInputEvent<HTMLInputElement>) => {
+    this.setState({performSubstitutions: evt.target.checked})
+  }
+
+  _onChangeSubstitutions = (evt: SyntheticInputEvent<HTMLInputElement>) => {
+    this.setState({substitutions: evt.target.value})
+  }
+
+  _onSaveTransformationSettings = () => {
+    this.props.onSave(this.state, this.props.index)
+  }
+
+  componentDidUpdate (prevProps: Props) {
+    const { transformation } = this.props
+    if (!isEqual(transformation, prevProps.transformation)) {
+      this.setState(getTransformationFields(transformation))
+    }
+  }
```

Comment:
This method isn't recommended according to a [React blog post](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key). To reset some values, one recommended option is to pass in a `key` prop that will be different whenever the transformation value is different.

Reply:
Good suggestion, updated in https://github.com/ibi-group/datatools-ui/pull/649/commits/a7296121781a1413670f4e5ad6eb4bafabbee7bb.